[B5] settings: API keys page

orangejenny opened the pull request:
## Product Description
This is the last view in settings.

It adds datepickers to our tempus dominus library, and it migrates the base CRUD page that's used in a few parts of HQ.

before
![Screenshot 2024-05-09 at 9 38 06 PM](https://github.com/dimagi/commcare-hq/assets/1486591/37089d93-84e7-41f9-874b-858dfbffd4df)

after
![Screenshot 2024-05-09 at 9 37 35 PM](https://github.com/dimagi/commcare-hq/assets/1486591/e9b1caed-f060-4c8d-8b54-4f21223bd790)


## Safety Assurance

### Safety story
UI changes, risk is pretty well limited to this page.

### Automated test coverage

Not much if any at the UI level.

### QA Plan

https://dimagi.atlassian.net/browse/QA-6513

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

orangejenny committed:
Updated markup in pagination.html

orangejenny committed:
Removed form-actions, which looks weird when the whole form is shaded

orangejenny committed:
Updated base_paginated_crud.html

orangejenny committed:
Updated hqwebapp/js/bootstrap5/crud_paginated_list

orangejenny committed:
Updated settings/js/user_api_keys

orangejenny committed:
Updated user_api_keys.html

orangejenny committed:
Switched view to B5

orangejenny committed:
Unsplit user_api_keys.html

orangejenny committed:
Unsplit settings/js/user_api_keys

orangejenny committed:
"Bootstrap 5 Migration - Rebuilt diffs"

orangejenny committed:
Merge branch 'master' into jls/b5-api-keys

jingcheng16 reviewed (commented):
Hi Jenny, I don't have enough knowledge to review the js changes you made in this PR. But left two questions for html changes.

orangejenny commented on corehq/apps/hqwebapp/templates/hqwebapp/bootstrap5/base_paginated_crud.html:
```diff
@@ -63,11 +62,7 @@
-                         style="margin: 0;">
-                      {{ pagination.text.empty }}
-                    </div>
+                    {{ pagination.text.empty }}
```
This is an intentional, though minor, visual change. Previously, if you didn't have any items, you saw a message in a blue bubble within the table, something like this (this exact screenshot is from a different page, but it's the same UI):

![Screenshot 2024-05-13 at 1 25 22 PM](https://github.com/dimagi/commcare-hq/assets/1486591/47f74e05-1eec-4aa3-9afa-4a231c731a06)

Now it's a plainer message:
![Screenshot 2024-05-13 at 1 25 38 PM](https://github.com/dimagi/commcare-hq/assets/1486591/3974a5df-9751-4c5b-8613-76b4237e9225)

orangejenny commented on corehq/apps/hqwebapp/templates/hqwebapp/bootstrap5/base_paginated_crud.html:
```diff
@@ -138,22 +133,16 @@
-              {% endfor %}
-            </select>
-          </div>
+          <select class="form-control" id="pagination-limit" data-bind="event: {change: updateListLimit}">
```
I've been doing it consistently, yes. The label isn't really adding value (and the inline form needed to align the text plus the select box properly isn't supported in B5).

orangejenny committed:
"Bootstrap 5 Migration - Marked 'settings' as complete"

millerdev commented:
Aside, and not a criticism of this PR: maybe it's a sign that I'm old, but I find the new inline labels and fields layout to be harder to read than the two-column layout on a large screen. I think it was changed to accommodate mobile devices, and if that's true I think it's a place where responsive design might be nice (use columns on large screen, inline on small).

biyeun commented:
@orangejenny @millerdev I think the form might be harder to follow now due to lack of whitespace. usually the fields should be wrapped in `mb-3` to add a margin at the bottom.

orangejenny committed:
Updated form classes

orangejenny commented:
@biyeun Good eye. https://github.com/dimagi/commcare-hq/pull/34592/ affected crispy rendering in some way such that the `mb-3` spacing wasn't being added. Fixed in f9e382a88b273758f9050c956548e16b79314a40, and the form does now have a bit more breathing room.
<img width="912" alt="Screenshot 2024-05-14 at 5 46 59 PM" src="https://github.com/dimagi/commcare-hq/assets/1486591/e35789ca-b761-4f2d-9fb4-b77f5dda0c14">

orangejenny committed:
Merge branch 'master' into jls/b5-api-keys

jingcheng16 commented:
Just curious, (if the two column layout refers to left column being label and right column being input field ),  why the two column layout becomes one column in this PR? I cannot identify the related code changes...

orangejenny commented:
@jingcheng16 It's almost invisible. Because the form on this page uses crispy forms, adding the `use_bootstrap5` decorator is enough to do the layout switch. Code like [this template](https://github.com/dimagi/crispy-bootstrap3to5/blob/main/crispy_bootstrap3to5/templates/bootstrap3to5/field.html) in our crispy3to5 repo checks the B5 flag to determine what kind of layout to display.

There are also places that key off the containing `form_class` CSS class, like [this](https://github.com/dimagi/crispy-bootstrap3to5/blob/f8b70d6bba1d6df7510561222ed04d430be19cbc/crispy_bootstrap3to5/templates/bootstrap3to5/bootstrap5/inputs.html#L2). Those made it necessary to make the change in https://github.com/dimagi/commcare-hq/pull/34593/commits/f9e382a88b273758f9050c956548e16b79314a40 which removes the `form-horizontal` class for B5 forms (the `crispy.py` change specifically, which overrides [this line](https://github.com/dimagi/commcare-hq/blob/f9e382a88b273758f9050c956548e16b79314a40/corehq/apps/hqwebapp/crispy.py#L24) earlier in `HQFormHelper`).

jingcheng16 commented:
@orangejenny I see.Thank you for the explanation and all the links! ❤️

orangejenny committed:
Restored createItemForm handler

This fires when the user submits the creation form, and django sends back
a new form, so the date picker needs to be re-initialized.
